### DISCUSSION TITLE
Noob question: Having a hard time understanding the native codec binary build steps.

### DISCUSSION THREAD

**Discussion by futuremotiondev** [saucecontrol/PhotoSauce#146] [Q&A]
I followed all of the steps:

```
git clone --recurse-submodules https://github.com/saucecontrol/PhotoSauce
git submodule update --init
git submodule update --remote
dotnet build src\MagicScaler -c Dist
```
And then the build succeeded:

```
Successfully created package 'D:\Dev\00 External Compile\PhotoSauce\build\..\out\nuget\PhotoSauce.MagicScaler.0.14.1.nupkg'.
Successfully created package 'D:\Dev\00 External Compile\PhotoSauce\build\..\out\nuget\PhotoSauce.MagicScaler.0.14.1.snupkg'.
```

But I'm now confused about this:

> Native codec binaries are built with [vcpkg](https://github.com/microsoft/vcpkg). See azure-pipelines.yml in the repo root for a working setup. Alternatively, pre-built native binaries can be retrieved from a [recent CI run](https://dev.azure.com/saucecontrol/PhotoSauce/_build?definitionId=1) and extracted to [reporoot]\out\vcpkg\install Once native binaries are built or downloaded, the native codec plugin packages can be built as normal with dotnet build

I downloaded all of the artifacts from here: 
https://dev.azure.com/saucecontrol/PhotoSauce/_artifacts/feed/photosauce_ci
And here:
https://dev.azure.com/saucecontrol/PhotoSauce/_artifacts/feed/photosauce_vcpkg

Am I supposed to place all of the `.nupkg` files in `PhotoSauce\out\vcpkg\install\` and then rebuild with `dotnet build`? Should I run `dotnet build src\MagicScaler -c Dist` again?

I just want to make sure that all of the native codec plugins are installed / working before I use the library.

Any guidance would be greatly appreciated!

**Comment by saucecontrol** [comment]
> I just want to make sure that all of the native codec plugins are installed / working before I use the library.

I'm a bit confused by this statement.  Everything you need to *use* the library (and plugins) is included in the packages published on [nuget.org](https://www.nuget.org/packages?q=photosauce).  In fact, if you intend to run on Windows, you probably don't even need any of the native codec packages; the main MagicScaler library includes bindings for the codecs built in to Windows.

The [CI feed](https://dev.azure.com/saucecontrol/PhotoSauce/_artifacts/feed/photosauce_ci) you linked includes the same packages, built from the latest commits.  These are also ready to use, as is.

The other feed you linked is for the [vcpkg cache](https://dev.azure.com/saucecontrol/PhotoSauce/_artifacts/feed/photosauce_vcpkg), which exists only to reduce build times in the event nothing changed in the native libraries or build environment between CI runs.  Those packages do contain the binaries, but they are only for use by vcpkg.  As the descriptions say, `NOT FOR DIRECT USE. Automatically generated cache package.`.

If you really do need to build the packages yourself (e.g. if you are making modifications to the code or building for a platform not included in the existing packages), you can save some build time by retrieving the expected `vcpkg/install` folder contents from a recent CI run.  If you click on any pipeline run in the list, you should see a link for published artifacts, which will let you download the install folder for any architecture run in CI.  
<img width="256" alt="image" src="https://github.com/saucecontrol/PhotoSauce/assets/7929102/c99bd0e0-20c6-48c2-8e1f-bd5cf2a65d67">

You can then build the codec packages individually using `dotnet build -c Dist src\NativeCodecs\Libjpeg`, etc.
